cylc/flow/ws_data_mgr.py

```diff
@@ -0,0 +1,393 @@
+                SUITE_STATUS_RUNNING_TO_STOP %
+                self.schd.stop_task)
+        elif self.schd.final_point:
+            status_string = (
+                SUITE_STATUS_RUNNING_TO_STOP %
+                self.schd.final_point)
+        else:
+            status_string = SUITE_STATUS_RUNNING
+        workflow.status = status_string
+
+        workflow.tasks.extend([t.id for t in tasks.values()])
+        workflow.families.extend([f.id for f in families.values()])
+
+        # Generate ungrouped edges
+        try:
+            graph_edges = config.get_graph_edges(
+                workflow.oldest_cycle_point,
+                workflow.newest_cycle_point,
+                group_nodes=None, ungroup_nodes=None,
+                ungroup_recursive=False, group_all=False
+            )
+        except TypeError:
+            graph_edges = []
+
+        if isinstance(graph_edges, list) and graph_edges != []:
+            for e_list in graph_edges:
+                if e_list[0] is None:
+                    continue
+                elif e_list[0] in task_proxies:
+                    t_node = task_proxies[e_list[0]].id
+                else:
+                    tn_name, tn_point = TaskID.split(e_list[0])
+                    t_node = f"{self.workflow_id}/{tn_point}/{tn_name}"
+                if e_list[1] is None:
+                    h_node = f"{self.workflow_id}/None"
+                elif e_list[1] in task_proxies:
+                    h_node = task_proxies[e_list[1]].id
+                else:
+                    hn_name, hn_point = TaskID.split(e_list[1])
+                    h_node = f"{self.workflow_id}/{hn_point}/{hn_name}"
+                e_id = e_list[0] + '/' + (e_list[1] or 'None')
+                edges[e_id] = PbEdge(
+                    id=f"{self.workflow_id}/{e_id}",
+                    tail_node=t_node,
+                    head_node=h_node,
+                    suicide=e_list[3],
+                    cond=e_list[4],
+                )
+            graph.edges.extend([e.id for e in edges.values()])
+            graph.leaves.extend(config.leaves)
+            graph.feet.extend(config.feet)
+            workflow.edges.CopyFrom(self.graph)
+            for key, info in config.suite_polling_tasks.items():
+                graph.workflow_polling_tasks.add(
+                    local_proxy=key,
+                    workflow=info[0],
+                    remote_proxy=info[1],
+                    req_state=info[2],
+                    graph_string=info[3],
+                )
+
+        # Replace the originals (atomic update, for access from other threads).
+        self.state_count_totals = state_count_totals
+        self.state_count_cycles = state_count_cycles
+        self.tasks = tasks
+        self.task_proxies = task_proxies
+        self.families = families
+        self.family_proxies = family_proxies
+        self.edges = edges
+        self.graph = graph
+        self.workflow = workflow
```

Comment:
Phew, finished reading this method. Kudos for writing it! I swear I had to get a piece of paper at some point to keep track of what was happening some of the loops. It was probably not easy to write this :clap: :clap: 

Before it gets merged, maybe even before the unit tests, I think we should break it into smaller methods. Each of these loops, and even other operations can be isolated.

I think this way it will be not only easier to maintain, but also easier for you to write the unit tests.

Reply:
Granulated